### PR TITLE
[Refactor] Jit-V2: evaluate row by row for uncompilable inputs

### PR DESCRIPTION
**Pull request by fzhedu**
Why I'm doing:

JIT V1 is not optimal for if clause, as V1 loads all inputs, not just as needed.

What I'm doing:

evaluate row by row for uncompilable inputs, benefits to if-clauses, like `casewhen SQL1`, besides, no degrade for normal cases. In detail, instead of loading all column data into LLVM values in advance, loading when hit the branch.

  | no-jit | jit-v1 | jit-v2
-- | -- | -- | --
SQL1 | 9.47 | 6.31 | 6.4
SQL2 | 3.22 | 2.9 | 2.91
SQL3 | 3.64 | 2.91 | 2.96
SQL4 | 9.57 | 7.42 | 7.36
casewhen SQL1 | 16.3 | 11.62 | 9.08

</byte-sheet-html-origin>

Fixes #issue
after https://github.com/StarRocks/starrocks/pull/40163

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/jit/jit_expr.cpp [hunk 1]
```diff
@@ -75,6 +75,9 @@ Status JITExpr::prepare(RuntimeState* state, ExprContext* context) {
     }
     if (_jit_function != nullptr) {
         _jit_expr_name = _expr->debug_string();
+        if (_jit_expr_name.empty()) {
+            return Status::RuntimeError("[JIT] expr debug_string() is empty");
+        }
```
Comment (silverbullet233) [inline be/src/exprs/jit/jit_expr.cpp:80]: it looks like a debug check, so using DCHECK here is better than returning an error?

File: be/src/exprs/jit/jit_engine.cpp [hunk 2]
```diff
@@ -123,6 +124,91 @@ StatusOr<JITScalarFunction> JITEngine::compile_scalar_function(ExprContext* cont
...
+    std::vector<LLVMColumn> columns(args_size + 1);
+
+    for (size_t i = 0; i < args_size + 1; ++i) {
+        // i == args_size is the result column.
+        auto* jit_column = b.CreateLoad(data_type, b.CreateConstInBoundsGEP1_64(data_type, columns_arg, i));
+
+        const auto& type = i == args_size ? expr->type() : input_exprs[i]->type();
+#if JIT_DEBUG
+        auto tmp = i == args_size ? expr : input_exprs[i];
+        LOG(INFO) << "[JIT] " << i << " col type = " << logical_type_to_string(type.type)
```
Comment (silverbullet233) [inline be/src/exprs/jit/jit_engine.cpp:163]: why not use VLOG，introducing a new macro seems not necessary?

File: be/src/exprs/jit/jit_expr.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exprs/jit/jit_expr.cpp:80]: debug_string() is replaced by jit_func_name() in later PRs, `jit_func_name()` will process this exception.

File: be/src/exprs/jit/jit_engine.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exprs/jit/jit_engine.cpp:163]: is removed in next PR.

File: be/src/exprs/jit/ir_helper.h [hunk 3]
```diff
@@ -50,6 +65,14 @@ struct LLVMColumn {
     llvm::Type* value_type = nullptr;  ///< Represents the type of the column's values.
 };
 
+struct JITContext {
+    llvm::Value* index_phi;
+    std::vector<LLVMColumn>& columns;
+    llvm::Module& module;
+    llvm::IRBuilder<>& builder;
+    int input_index;
```
Comment (silverbullet233) [inline be/src/exprs/jit/ir_helper.h:73]: Initialization is required here

Comment (fzhedu) [reply be/src/exprs/jit/ir_helper.h:73]: just used once with values. I will init it to 0 in next PR. https://github.com/StarRocks/starrocks/pull/40166/files

**Review by silverbullet233 — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**